lib/browse_everything/driver/sharepoint.rb

```diff
@@ -0,0 +1,232 @@
+          values << directory_entry(f)
+        end
+        @entries = values.compact
+
+        @sorter.call(@entries)
+      end
+
+      # @return [String]
+      # Authorization url that is used to request the initial access code from Sharepoint/Onedrive/365/etc
+      def auth_link(*_args)
+        Addressable::URI.parse("https://login.microsoftonline.com/#{config[:tenant_id]}/oauth2/v2.0/authorize?#{auth_query_string}")
+      end
+
+      # @return [Boolean]
+      def authorized?
+        @token.present?
+      end
+
+      def authorize!
+        return if @code.blank?
+        register_access_token(sharepoint_session.get_access_token(@code))
+        @code = nil
+        @token
+      end
+
+      def connect(params, _data, _url_options)
+        @code = params[:code]
+        authorize!
+      end
+
+      # @param [String] id of the file on MS graph drive
+      # @return [Array<String, Hash>]
+      def link_for(id)
+        file = items_by_id(id)
+        extras = {file_name: file['name'], file_size: file['size'].to_i}
+        [download_url(file), extras]
+      end
+
+      private
+
+      def auth_query_string
+        query = []
+        # keep_if deletes from and returns self, so dup config to not overwrite original
+        base = config.dup.keep_if { |k,v| ['client_id', 'scope', 'redirect_uri'].include?(k) }
+        base.each do |k,v|
+          query += ["#{k}=#{v}"]
+        end
+        query += ["response_type=code"]
+
+        query.join('&')
+      end
+
+      def session
+        AuthenticationFactory.new(
+          self.class.authentication_klass,
+          client_id: config[:client_id],
+          client_secret: config[:client_secret],
+          tenant_id: config[:tenant_id],
+          scope: config[:scope],
+          redirect_uri: config[:redirect_uri],
+          code: @code.presence,
+          access_token: @token.presence
+        )
+      end
+
+      def authenticate
+        session.authenticate
+      end
+
+      def sharepoint_session
+        @sharepoint_session ||= authenticate
+      end
+
+      def token_refresh
+        return @token unless token_expired?
+
+        register_access_token(sharepoint_session.refresh_token)
+      end
+
+      # If there is an active session, {@token} will be set by {BrowseEverythingController} using data stored in the
+      # session. 
+      #
+      # @param [OAuth2::AccessToken] access_token
+      def register_access_token(access_token)
+        @token = {
+                   'token' => access_token.token,
+                   'expires_in' => access_token.expires_in,
+                   'expires_at' => access_token.expires_at,
+                   'refresh_token' => access_token.refresh_token
+                 }
+      end
+
+      def sharepoint_token
+        return unless @token
+        @token.fetch('token', nil)
+      end
+
+      def expiration_time
+        return unless @token
+        @token.fetch('expires_at', nil).to_i
```

Comment:
I think you'll need to add special handling for nil because `nil.to_i == 0` which would cause the `expiration_time.nil?` check in `.token_expired?` fail.  Maybe something like this?
```suggestion
        expires_at = @token.fetch('expires_at', nil)
        expires_at.nil? ? nil : expires_at.to_i
```